PR - Journalist can create article

Ehugo2000 opened the pull request:
[PT-story](https://www.pivotaltracker.com/story/show/176041267)
### User Story
```
As an API,
In order to keep a record of content,
I would like to be able to save a created article in a database.
```
## Changes proposed in this pull request:
* Adding Create action to articles_controller
* Creating category model
* Adding category in seeds.rb
## What I have learned working on this feature:
* Working with Postman
* Params
* Debugging
* Creating an article

FlisAnn committed:
Creating test for journalist can create article

FlisAnn committed:
Adding route and create action

FlisAnn committed:
Creates article and return 201

FlisAnn committed:
article returns a success message

FlisAnn committed:
Adding private method for article_params

Ehugo2000 committed:
Adding sad path in test and in controller for create action.

emtalen commented on app/controllers/api/articles_controller.rb:
```diff
@@ -10,4 +10,20 @@ def show
+    if article.persisted?
+      render json: { message: 'Your article was successfully created' }, status: 201
+    else
+      render json: { message: 'Something went wrong!' }, status: 422
```
use the active record errors that you get if a title isn't added and so on. Put a debugger and get your way through to the specific errors.

emtalen commented on spec/requests/api/journalist_can_create_an_article.spec.rb:
```diff
@@ -0,0 +1,43 @@
+RSpec.describe 'POST/api/articles', type: :request do
+  let!(:articles) { create(:article) }
```
why are you creating an article?

Ehugo2000 replied:
Dont we need it to create an Article in the test?

Ehugo2000 replied:
it is removed now btw

emtalen replied:
@Ehugo2000 only if you want to return an article, in this case you want to create it.

emtalen commented on spec/requests/api/journalist_can_create_an_article.spec.rb:
```diff
@@ -0,0 +1,43 @@
+      expect(response_json['message']).to eq 'Something went wrong!'
+    end
+  end
+end
```
you need to rename this file to _spec, that's why coveralls doesn't find it

Ehugo2000 replied:
Fixed

Ehugo2000 committed:
Updated articles_controller with a dynamic error message.
Updating sad path test for unsuccessfull creating an article.

Ehugo2000 committed:
Fixed merge conflict in routes.

Ehugo2000 committed:
Merge branch 'development' of github.com:Ehugo2000/kitty_news_api into journalist_can_create_article

ViktorHek committed:
merging uppdated version. fixing conflict in routes

ViktorHek committed:
adding category_id to article_params. added category to journalist_can_create_an_article_spec.rb

ViktorHek committed:
adding sad path tast. for atricle without category_id

KyunginNa committed:
Formatting code
Deleting enum in category model, changing label to string
Adding categories in seeds.rb

Ehugo2000 committed:
Fix merge conflict in routes

ViktorHek commented on config/routes.rb:
```diff
@@ -1,6 +1,6 @@
 Rails.application.routes.draw do
   namespace :api do
-    resources :articles, only: %i[index show]
+    resources :articles, only: %i[index show create]
```
I think we have to change it to ```only: [:index, :show, :create]```. this works but the other way looks more readable.

Ehugo2000 replied:
Fixed!

KyunginNa commented on spec/requests/api/journalist_can_create_an_article_spec.rb:
```diff
@@ -0,0 +1,91 @@
+               body: 'With a body',
+               category_id: category.id
+             },
+
```
We can delete this additional line:)

Ehugo2000 replied:
Fixed!

Ehugo2000 committed:
Resolve code reviews from Viktor and Kyungin.

emtalen commented on spec/requests/api/journalist_can_create_an_article_spec.rb:
```diff
@@ -0,0 +1,90 @@
+               title: 'A title',
+               lead: 'And a lead',
+               body: 'With a body',
+               category_id: category.id
```
How will you get the id of the category in the client? Would it be an easier solution for you in the client to sned in the label of the category and then here convert that by finding the category with that label and pass in that id when you save the article?
Just so you have thought this through and know how to deal with it in the client. This is a questions so I want an answer, there are no right or wrongs.

ViktorHek replied:
we are sending category_id from client

ViktorHek committed:
runnig seed and migration

ViktorHek committed:
params to find category is changed from label to id

ViktorHek committed:
refactor

ViktorHek committed:
refactor again

ViktorHek committed:
fixing mege conflict

ViktorHek committed:
remove id: from find(params['id']). modify visitor_can_view_article_by_category_spec

Ehugo2000 reviewed (commented):
Looks good team!
